### PR TITLE
fix: `Bun.deepMatch` on circular objects causing segfault

### LINKED ISSUE #15671: `Bun.deepMatch(a, b)` segfaults on objects with circular references.
### What version of Bun is running?

1.1.39-debug+58d203367

### What platform is your computer?

Darwin 24.1.0 arm64 arm

### What steps can reproduce the bug?

```ts
describe("Bun.deepMatch(a, b)", () => {
  it("handles circular references", () => {
    let foo = { bar: undefined as any };
    let bar = { foo: undefined as any };
    foo.bar = bar;
    bar.foo = foo;
    expect(Bun.deepMatch(foo, foo)).toBe(true);
  });
});
```

### What is the expected behavior?

This test should pass

### What do you see instead?

```
Bun Debug v1.1.39 (58d20336) macOS Silicon
macOS v15.1.1
Args: "bun-debug" "test" "test/js/bun/bun-object"
Features: bunfig jsc tsconfig(3) tsconfig_paths
Builtins: "bun:jsc" "detect-libc" "node:child_process" "node:fs" "node:fs/promises" "node:os" "node:path" "node:string_decoder" "node:util/types"
Elapsed: 120ms | User: 124ms | Sys: 35ms
RSS: 0.15GB | Peak: 0.15GB | Commit: 1.07GB | Faults: 124

panic(main thread): Segmentation fault at address 0x16CA33FF8
???:?:?: 0x18d5dbe23 in ??? (libsystem_platform.dylib)
???:?:?: 0x30130001022ee577 in ??? (???)
cache/webkit-8f9ae4f01a047c666ef548864294e01df731d4ea/include/wtf/Vector.h:487:11: 0x1022ee577 in VectorBuffer (../../src/bun.js/bindings/BunInjectedScriptHost.cpp)
cache/webkit-8f9ae4f01a047c666ef548864294e01df731d4ea/include/wtf/Vector.h:693:5: 0x1022ee4ef in Vector (../../src/bun.js/bindings/BunInjectedScriptHost.cpp)
cache/webkit-8f9ae4f01a047c666ef548864294e01df731d4ea/include/wtf/Vector.h:694:5: 0x1022ee40b in Vector (../../src/bun.js/bindings/BunInjectedScriptHost.cpp)
cache/webkit-8f9ae4f01a047c666ef548864294e01df731d4ea/include/JavaScriptCore/PropertyNameArray.h:39:5: 0x1022ee34b in PropertyNameArrayData (../../src/bun.js/bindings/BunInjectedScriptHost.cpp)
cache/webkit-8f9ae4f01a047c666ef548864294e01df731d4ea/include/JavaScriptCore/PropertyNameArray.h:40:5: 0x1022ee20f in PropertyNameArrayData (../../src/bun.js/bindings/BunInjectedScriptHost.cpp)
cache/webkit-8f9ae4f01a047c666ef548864294e01df731d4ea/include/JavaScriptCore/PropertyNameArray.h:34:71: 0x1022edfe7 in create (../../src/bun.js/bindings/BunInjectedScriptHost.cpp)
cache/webkit-8f9ae4f01a047c666ef548864294e01df731d4ea/include/JavaScriptCore/PropertyNameArray.h:50:18: 0x1022edf33 in PropertyNameArray (../../src/bun.js/bindings/BunInjectedScriptHost.cpp)
cache/webkit-8f9ae4f01a047c666ef548864294e01df731d4ea/include/JavaScriptCore/PropertyNameArray.h:54:5: 0x1022edad3 in PropertyNameArray (../../src/bun.js/bindings/BunInjectedScriptHost.cpp)
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by DonIsaac**
### What does this PR do?
Fixes #15671
- Fixes a segfault in `Bun.deepMatch(a, b)` when either `a` or `b` have a circular reference
- Adds a `==` short circuit to `Bun__deepMatch` for referentially-equal objects and properties.

This segfault was appearing in bun's own test suite. Part of the effort to get CI green.

### How did you verify your code works?

Added JS test cases.

### CODE REVIEW

**Comment by robobun** [comment]
Your commit 7f5461335e7d2077c456e894555bac60c4885c58 is building: [`#7654`](https://buildkite.com/bun/bun/builds/7654)

File: src/bun.js/bindings/bindings.cpp [hunk 1]
```diff
@@ -1354,9 +1357,42 @@ bool Bun__deepEquals(JSC__JSGlobalObject* globalObject, JSValue v1, JSValue v2,
...
+ *
+ * @return true
+ * @return false
+ */
 template<bool enableAsymmetricMatchers>
-bool Bun__deepMatch(JSValue objValue, JSValue subsetValue, JSGlobalObject* globalObject, ThrowScope* throwScope, bool replacePropsWithAsymmetricMatchers, bool isMatchingObjectContaining)
-{
+bool Bun__deepMatch(
+    JSValue objValue,
+    std::set<EncodedJSValue>* seenObjProperties,
```
Comment (dylan-conway) [inline src/bun.js/bindings/bindings.cpp:1390]: Another argument to hold `JSValue`s added to `seenObjProperties` and `seenSubsetProperties` should be added. Something like
```cpp
MarkedArgumentBuffer& gcBuffer
```

File: src/bun.js/bindings/bindings.cpp [hunk 2]
```diff
@@ -1354,9 +1357,42 @@ bool Bun__deepEquals(JSC__JSGlobalObject* globalObject, JSValue v1, JSValue v2,
...
+    JSValue subsetValue,
+    std::set<EncodedJSValue>* seenSubsetProperties,
+    JSGlobalObject* globalObject,
+    ThrowScope* throwScope,
+    bool replacePropsWithAsymmetricMatchers,
+    bool isMatchingObjectContaining)
+{
+
+    // fast path for reference equality.
+    if (objValue == subsetValue) return true;
```
Comment (dylan-conway) [inline src/bun.js/bindings/bindings.cpp:1404]: also an assertion should be added
```c++
ASSERT(objValue.isCell());
ASSERT(subsetValue.isCell());
```

Comment (DonIsaac) [reply src/bun.js/bindings/bindings.cpp:1404]: Added

File: src/bun.js/bindings/bindings.cpp (hunk 1 above)
Comment (DonIsaac) [reply src/bun.js/bindings/bindings.cpp:1390]: Done

**Review by dylan-conway — APPROVED** [review]
Looks good. Let's merge after ci

**Pull request merged**